github/commitstatus_private_test.go

```diff
@@ -0,0 +1,80 @@
+package github
+
+import (
+	"net/http"
+	"testing"
+	"time"
+
+	"gotest.tools/v3/assert"
+)
+
+func TestCheckForRetrySuccess(t *testing.T) {
+	type testCase struct {
+		name         string
+		res          httpResponse
+		waitTime     time.Duration
+		maxSleepTime time.Duration
+		jitter       time.Duration
+		wantSleep    time.Duration
+		wantReason   string
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		sleep, reason, err := checkForRetry(tc.res, tc.waitTime, tc.maxSleepTime, tc.jitter)
+
+		assert.NilError(t, err)
+		assert.Equal(t, sleep, tc.wantSleep)
+		assert.Equal(t, reason, tc.wantReason)
+	}
+
+	testCases := []testCase{
+		{
+			name:      "status OK: sleep==0",
+			res:       httpResponse{statusCode: http.StatusOK},
+			wantSleep: 0 * time.Second,
+		},
+		{
+			name:      "non retryable status code: sleep==0",
+			res:       httpResponse{statusCode: http.StatusTeapot},
+			wantSleep: 0 * time.Second,
+		},
+		{
+			name:       "retryable status code: sleep==waitTime",
+			res:        httpResponse{statusCode: http.StatusInternalServerError},
+			waitTime:   42 * time.Second,
+			wantSleep:  42 * time.Second,
+			wantReason: "Internal Server Error",
+		},
+		{
+			name: "rate limited, would sleep too long: sleep==0",
+			res: httpResponse{
+				statusCode: http.StatusForbidden,
+				date:       time.Date(2001, time.April, 30, 13, 0, 0, 0, time.UTC),
+				// 1h after
+				rateLimitReset: time.Date(2001, time.April, 30, 14, 0, 0, 0, time.UTC),
+			},
+			maxSleepTime: 10 * time.Second,
+			wantSleep:    0 * time.Second,
+		},
+		{
+			name: "rate limited, would sleep a bit, adding also the jitter",
+			res: httpResponse{
+				statusCode: http.StatusForbidden,
+				date:       time.Date(2001, time.April, 30, 13, 0, 0, 0, time.UTC),
+				// 1h after
+				rateLimitReset: time.Date(2001, time.April, 30, 14, 0, 0, 0, time.UTC),
```

Comment:
this works but it doesn't really make sense... Rate limit is reset each hour, so I would keep this value to <1h.
and also set `maxSleepTime` to a value smaller than 1h...

The code supports any values, that is the reason your tests passes, but this is how the server (Github) works so I would like to be in line with reality. It will be easier to understand later no?

Reply:
The reason I wrote it in this way is exactly to avoid letting the reader thinking that the code knows anything about the GitHub time constants.

I understand it can be confusing; let me think about it.

Reply:
Updated.

Reply:
Thanks.